### PR TITLE
JSON consist format: https://blueprints.launchpad.net/or/+spec/consist-editor

### PR DESCRIPTION
**Pull request by YoRyan**
Implements a native JSON format for consists, including a new loader, consolidated consist spawning code, updates to the content checkers, and GUI changes in the launcher. A detailed specification is available on [Elvas Tower](http://www.elvastower.com/forums/index.php?/topic/34187-or-consist-format/).

### CODE REVIEW

**Comment by YoRyan** [comment]
This PR is now ready for review.

**Comment by YoRyan** [comment]
I am still not convinced that "Block" is the right term, but I do agree it would be nice to get rid of the ambiguity. Therefore, I will rename "Consist" classes to "WagonList".

**Comment by YoRyan** [comment]
Correction, I have adopted the "Train" name for the new classes and format. See the [discussion](http://www.elvastower.com/forums/index.php?/topic/34187-or-consist-format/page__view__findpost__p__260817) on Elvas Tower.

**Comment by YoRyan** [comment]
So, to sum up: The current naming scheme reflects [my position on the matter](http://www.elvastower.com/forums/index.php?/topic/34187-or-consist-format/page__view__findpost__p__260888) on Elvas Tower.

If there are no further objections, I believe this PR is ready for a second look and merge. I also written some tests, but I'll hold off on those until a decision is made on #234.

File: Source/ORTS.Menu/Consists.cs [hunk 1]
```diff
@@ -56,116 +61,98 @@ internal Consist(string filePath, Folder folder)
...
-            FilePath = filePath;
         }
 
         public override string ToString()
         {
             return Name;
         }
 
-        public static List<Consist> GetConsists(Folder folder)
+        public static List<Consist> GetConsists(Folder folder, IList<Folder> allFolders)
```
Comment (twpol) [inline Source/ORTS.Menu/Consists.cs:82]: I really am feeling like the work necessary for the "profile" property is excessive, for something that will have to be 100% replaced in the near future.

File: Source/Menu/TestingForm.cs [hunk 2]
```diff
@@ -111,9 +111,10 @@ void LoadActivities()
 
             TestActivityLoader = new Task<SortableBindingList<TestActivity>>(this, () =>
             {
-                return new SortableBindingList<TestActivity>((from f in Folder.GetFolders(Settings)
+                IList<Folder> folders = Folder.GetFolders(Settings);
```
[the remark this answers is not in the document]
Comment (YoRyan) [reply Source/Menu/TestingForm.cs:114]: Very well. I personally follow the MSDN [convention](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions#implicitly-typed-local-variables) of declaring types if they aren't already present in the right-hand side.

File: Source/ORTS.Common/ORTS.Common.csproj [hunk 3]
```diff
@@ -53,6 +53,7 @@
     <Compile Include="AlmostEqual.cs" />
     <Compile Include="CancellationToken.cs" />
     <Compile Include="CancellationTokenSource.cs" />
+    <Compile Include="TrainFile.cs" />
```
[the remark this answers is not in the document]
Comment (YoRyan) [reply Source/ORTS.Common/ORTS.Common.csproj:56]: Sure, I can move it.

File: Source/Menu/MainForm.cs [hunk 4]
```diff
@@ -891,7 +902,7 @@ void ShowLocomotiveList()
             {
                 var consist = SelectedActivity.Consist;
                 comboBoxLocomotive.Items.Clear();
-                comboBoxLocomotive.Items.Add(consist.Locomotive);
+                comboBoxLocomotive.Items.AddRange(consist.Locomotives.ToArray());
```
[the remark this answers is not in the document]
Comment (YoRyan) [reply Source/Menu/MainForm.cs:905]: Some folks on Elvas Tower have been helping me resolve some of them. I'll try to put the menu through a battery of tests myself.

File: Source/Menu/MainForm.Designer.cs [hunk 5]
```diff
@@ -926,11 +978,53 @@ private void InitializeComponent()
...
+            this.comboBoxTimetableLocomotive.Size = new System.Drawing.Size(256, 24);
+            this.comboBoxTimetableLocomotive.TabIndex = 9;
+            // 
             // MainForm
             // 
-            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
+            this.AutoScaleDimensions = new System.Drawing.SizeF(8F, 16F);
             this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
-            this.ClientSize = new System.Drawing.Size(854, 548);
+            this.ClientSize = new System.Drawing.Size(1139, 674);
```
[the remark this answers is not in the document]
Comment (YoRyan) [reply Source/Menu/MainForm.Designer.cs:1027]: Hmm, I will see if I can avoid the resize.

File: Source/ORTS.Menu/Consists.cs (hunk 1 above)
Comment (YoRyan) [reply Source/ORTS.Menu/Consists.cs:82]: If we're certain that profiles are absolutely not the way to go, I would be open to dropping this feature and reverting these changes.

File: Source/Menu/MainForm.Designer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (YoRyan) [reply Source/Menu/MainForm.Designer.cs]: Issue was caused by my use of 125% font scaling on my hi-DPI monitor.

**Comment by YoRyan** [comment]
.train-or files have been moved back to the CONSISTS folder per the Elvas Tower [discussion](http://www.elvastower.com/forums/index.php?/topic/34187-or-consist-format/page__view__findpost__p__260907).

**Comment by Sharpe49** [comment]
Sorry for previous message, but I had weird notification messages.

**Comment by YoRyan** [comment]
Will do. I will of course cover how .train-or files substitute for .con ones in the CONSISTS folder, but should I also include a specification of the new format? I intend to start work on an editor in the near-future, so hopefully users won't have to write their own train files.

**Comment by YoRyan** [comment]
I have added some manual documentation, as well as a new "IsTilting" flag to represent tilting trains.

**Comment by YoRyan** [comment]
Hi Chris,

I thought "IVehicleList" would be a good generic descriptor for internal use only, so that's why I made the name change. The "train-or" name would be the one that is user-facing; as for what it should be, I'm still undecided, and I'm even starting to wonder if "consist" wasn't such a bad name after all.

If the management team can agree on a name, I will defer to that decision.

**Comment by YoRyan** [comment]
Removing from the unstable build as a Tests.csproj change conflicts with #230.

**Pull request closed without merging**